### PR TITLE
GH-95861: Add support for Spearman's rank correlation coefficient

### PR DESCRIPTION
**Pull request by rhettinger**
Still need to think about whether `by_rank` is the best name for the flag.

* Issue: gh-95861

### CODE REVIEW

**Comment by stevendaprano** [comment]
See my early comment on #95861 

I think we should make `_rank` public, and remove the "by_rank" parameter to correlation, moving that functionality into a thin wrapper function `spearman`. Otherwise this is good work, thank you.

File: Lib/statistics.py [hunk 1]
```diff
@@ -355,6 +355,31 @@ def _fail_neg(values, errmsg='negative value'):
...
+def _rank(data, /) -> list[float]:
+    """Rank order a dataset. The lowest value has rank 1.
+
+    Ties are averaged so that equal values receive the same rank.
+    The operation is idempotent.
+
+    >>> data = [31, 56, 31, 25, 75, 18]
+    >>> _rank(data)
+    [3.5, 5.0, 3.5, 2.0, 6.0, 1.0]
+    >>> _rank(_)
```
Comment (stevendaprano) [inline Lib/statistics.py]: Is the special behaviour of `_` guaranteed by doctest? I suppose this is safe enough.

**Review by stevendaprano — CHANGES_REQUESTED** [review]
If you categorically disagree with my suggestion to make rank public and move spearman into its own named function, I will accept this (you've done the hard work writing the code and I do like the feature). But I do strongly request that we keep to the design of separate named functions rather than overloading the one function with a flag.

If you categorically want to keep the single correlation function, then we could future-proof it and make the type of correlation more explicit with an enumerated flag rather than a bool:

```
def correlation(x, y, /, kind='pearson'):
    if kind == 'spearman':
        x = _rank(x)
        y = _rank(y)
   elif kind != 'pearson':
       raise ValueError('unsupported correlation kind')
```

If we ever add another sort of correlation (say, Kendall tau) is would be awkward to handle a second bool flag.

File: Lib/statistics.py (hunk 1 above)
Comment (rhettinger) [reply Lib/statistics.py]: Yes.  However, it is easy to remove so I just took it out.

**Comment by rhettinger** [comment]
Thank you for responding quickly and thoughtfully.

I added some comments to the issue to help us get to a good meeting of the minds on the API.

**Pull request merged**